onnx_chainer/replace_func.py

```diff
@@ -0,0 +1,208 @@
+import inspect
+
+import chainer
+
+
+class WrappedFunctionNode(chainer.FunctionNode):
+    """Wrap the target function and operate as ``FunctionNode``
+
+    Arguments:
+        name (str): name of the function node
+        func (func): the target function
+        args (str): args for the function
+        kwargs (str): kwargs for the function
+        attributes (list): parameters to be set node's attributes
+    """
+
+    def __init__(self, name, func, args, kwargs, attributes=None):
+        self.user_name = name
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+
+        if attributes is not None:
+            for k, v in attributes.items():
+                setattr(self, k, v)
+
+    def forward(self, xs):
+        self.xs = xs
+        results = self.func(*self.args, **self.kwargs)
+        if isinstance(results, (tuple, list)):
+            dummy_results = tuple(_unwrap_var(ret) for ret in results)
+        elif isinstance(results, dict):
+            dummy_results = tuple(_unwrap_var(ret) for ret in results.values())
+        else:
+            dummy_results = _unwrap_var(results)
+            dummy_results = dummy_results,
+        if not chainer.is_arrays_compatible(dummy_results):
+            raise ValueError(
+                'returned values from the function wrapped by \'as_funcnode\' '
+                'must consist only array, function name: {}'.format(self.name))
+        return dummy_results
+
+    def backward(self, indexes, gys):
+        xp = chainer.backend.get_array_module(self.xs[0])
+        ret = tuple(chainer.Variable(xp.zeros_like(x)) for x in self.xs)
+        return ret
+
+
+def fake_as_funcnode(alt_func, name, rename_attributes=None):
+    """The target function fakes FunctionNode
+
+    The target function is replaced to the alternative function to connect
+    variable node by acting function node. ``alt_func`` must satisfy the
+    following restrictions.
+
+    1. Inputs includes one or more ``chainer.Variable`` to trace variables.
+    2. Output consists nothing but ``ndarray`` or ``chainer.Variable``
+
+    Even if ``alt_func`` returns ``ndarray``, the value forced to be converted
+    to ``chainer.Variable``. A caller of the target function have to care
+    both cases, returning ``ndarray`` and ``chainer.Variable``.
+
+    When ``alt_func`` returns ``list`` of variable, the wrapped function will
+    also returns multiple variables as ``tuple``. However ``dict`` cannot
+    be return, the wrapped function breaks down the returned values as
+    ``tuple`` of values, keys will be ignored.
+
+    Arguments of ``alt_func`` except for ``chainer.Variable`` are set as
+    function attributes. Attribute names are set ``argN`` (N is index
+    number) or keyword on default.
+
+    Example::
```

Comment:
s/::/:/ ?

Reply:
``::`` means Sphinx's literal directive, and `::` is correct, but other docs do not follow sphinx style, so I'll deal with documentation fixing later. So I applied `:` on this time.